torch/csrc/jit/codegen/cuda/manager.cpp

```diff
@@ -86,42 +76,42 @@ class CudaFusionManager {
     TORCH_CHECK(
         kernel_cache_.count(kernel_id) != 0, "kernel id not recognized");
 
-    // TODO: temporary hack
-    auto cuda_kernel = kernel_cache_[kernel_id].getKernelPtr(inputs);
-    if (cuda_kernel) {
+    if (auto cuda_kernel_opt =
+            kernel_cache_[kernel_id].getKernelPtr(inputs, broadcasted_shape)) {
       // TODO: update launch config for specific sizes;
       //       maybe we should store it in CudaKernel and compute it later
-      runKernel(*cuda_kernel, inputs, outputs, broadcasted_shape);
+      runKernel(*cuda_kernel_opt, inputs, outputs, broadcasted_shape);
     } else {
       // TODO: this should somehow be done after kernel compilation.
       //       we will want compileKernel to return a heuristic
-      cuda_kernel = kernel_cache_[kernel_id].allocateKernelInCache(
-          makePWKernelSupport(inputs));
+      auto cuda_kernel = kernel_cache_[kernel_id].allocateKernelInCache(inputs);
 
       // lower torch::jit::Graph to torch::jit::fuser::cuda::fusion
       // TODO: pass contiguity infor as well as size req, so we can apply proper
       //       transform to computation
-      // we should propagate more information back:
-      //   1. device;
-      //   2. launch config;
-      parseJitIR(graph, cuda_kernel.value());
-      scheduleFusion(cuda_kernel.value()->fusion_.get(), inputs);
+      cuda_kernel->setFusionPtr(parseJitIR(graph));
+      TORCH_INTERNAL_ASSERT(
+          cuda_kernel->getFusionPtr() != nullptr,
+          "parser failed to construct a fusion from PyTorch JIT graph\n");
+
+      // magic scheduler
+      scheduleFusion(cuda_kernel->getFusionPtr(), inputs);
```

Comment:
add a comment reminding that the fusion will be mutated here?
(I'd actually prefer a different interface, where scheduleFusion() returns the new Fusion, but that may be outside the scope of this PR)

Reply:
Agreed.